### PR TITLE
fix: preserve VK location-only inbound messages

### PR DESCRIPTION
**Pull request by sirmax672**
## Summary
- preserve VK location-only inbound messages instead of dropping them as empty bodies
- resolve geo from multiple VK payload layers and parse both object and `"lon lat"` string coordinate formats
- enrich missing geo via VK API/history fallback and add focused tests for geo parsing and monitor handling

## Problem
In production, VK messages sent as native location attachments were reaching OpenClaw without a usable visible body. Two cases were affected:
- a message containing only a VK location could be ignored because the effective inbound body became empty
- a message with text plus VK location lost the location context before it reached the agent

This happened because `vk-io` payload hydration is inconsistent for `geo`: sometimes it is present in raw message payloads, sometimes only recoverable via follow-up API calls, and raw coordinates can arrive as a string in `lon lat` order.

## What changed
- added `VkInboundGeo` to the inbound message model
- taught media helpers to parse and format VK geo payloads
- made inbound body generation include a `[VK location] lat, lon` line when geo is present
- taught the monitor to look for geo in multiple context layers and to fall back to `messages.getById` / `messages.getByConversationMessageId` / `messages.getHistory` when needed
- added tests for geo parsing, location-only message rendering, and history-based geo recovery

## Testing
- `npm test`

### CODE REVIEW

File: src/monitor.ts [hunk 1]
```diff
@@ -98,34 +231,102 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
+        } catch (err) {
+          const errorMessage = err instanceof Error ? err.message : String(err);
+          opts.runtime.log?.(
+            `vk: failed to hydrate message payload for peerId=${context.peerId} messageId=${context.id}: ${errorMessage}`,
+          );
+        }
+      }
+    }
+
+    if (!fetchedMessage?.geo) {
```
Comment (pfrankov) [inline src/monitor.ts]: This fallback currently runs even when the incoming message has no geo/location signal. For an ordinary text-only message, `shouldEnrichFromApi(context)` is true when the context is not filled or has no geo, and then this block calls `messages.getHistory` whenever the direct fetch has no `geo`. That means normal inbound messages can pay one or two VK API calls before dispatch, increasing latency and rate-limit pressure. Please gate the direct/history enrichment to messages that actually have a geo signal, and add a regression test that a regular text message does not call `messages.getById` or `messages.getHistory`.

File: src/monitor.ts [hunk 2]
```diff
@@ -98,34 +231,102 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
-          ? context.conversationMessageId
-          : undefined,
       peerId,
       senderId,
-      text,
+      text: visibleBody,
       timestamp: createdAtSeconds ? createdAtSeconds * 1000 : Date.now(),
       isGroup,
-      messagePayload: context.messagePayload,
+      messagePayload: resolveVkMessagePayload(fetchedMessage?.payload ?? context.messagePayload),
```
Comment (pfrankov) [inline src/monitor.ts:335]: The constructed `VkInboundMessage` no longer includes `conversationMessageId`. The previous monitor path preserved it, and `handleVkInbound` uses it to enable status reactions (`createVkStatusReactionController` needs the VK cmid). With this PR, messages that should show queued/thinking/done reactions will silently skip that flow. Please preserve `conversationMessageId` from `context.conversationMessageId` or `fetchedMessage.conversation_message_id`, and cover it with a monitor test.

**Review by pfrankov — CHANGES_REQUESTED** [review]
Thanks for fixing the location-only message path. I do not think this is ready to merge yet: the current monitor changes add extra VK API calls to the normal inbound path and also drop `conversationMessageId`, which disables the existing status-reaction flow.

**Comment by sirmax672** [comment]
Addressed both review points:

- geo enrichment is now gated behind an actual geo signal, so ordinary text-only inbound messages no longer call `messages.getById` / `messages.getHistory`
- `conversationMessageId` is preserved again for downstream status reactions
- added monitor regressions for both cases

Re-ran tests after the update: `npm test` (`541 passed`).

File: src/monitor.ts [hunk 3]
```diff
@@ -18,6 +20,146 @@ export type VkMonitorOptions = {
...
+        normalizedText &&
+        item.text?.trim() === normalizedText &&
+        item.from_id === params.context.senderId
+      ) {
+        return true;
+      }
+      if (
+        typeof item.date === "number" &&
+        typeof params.context.createdAt === "number" &&
+        Math.abs(item.date - params.context.createdAt) <= 120 &&
```
Comment (pfrankov) [inline src/monitor.ts:147]: Please remove this fuzzy time/sender match. In a busy dialog, another message from the same sender within two minutes can be selected and its coordinates attributed to the current turn. Location data must be resolved only by exact message ID or exact peer + conversation message ID; otherwise fail closed.

File: src/monitor.ts [hunk 4]
```diff
@@ -98,12 +240,78 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
+        } catch (err) {
+          const errorMessage = err instanceof Error ? err.message : String(err);
+          opts.runtime.log?.(
+            `vk: failed to hydrate message payload for peerId=${context.peerId} messageId=${context.id}: ${errorMessage}`,
+          );
+        }
+      }
+    }
+
+    if (geoSignalPresent && !fetchedMessage?.geo) {
```
Comment (pfrankov) [inline src/monitor.ts:267]: After `context.loadMessagePayload()` succeeds, it mutates the context but leaves `fetchedMessage` undefined. This condition therefore still runs `getHistory` after successful hydration. Parse the raw geo first, hydrate once only when needed, then re-read the context before considering any further fallback.

File: src/monitor.ts [hunk 5]
```diff
@@ -98,12 +240,78 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
+    }
+
     const peerId = context.peerId;
     const senderId = context.senderId;
-    const text = context.text ?? "";
+    const text = fetchedMessage?.text ?? context.text ?? "";
     const isGroup = peerId >= 2_000_000_000;
-    const attachments = extractVkInboundAttachments(context.attachments);
-    const replyContext = resolveVkInboundReplyContext(context.replyMessage);
+    const attachments = extractVkInboundAttachments(fetchedMessage?.attachments ?? context.attachments);
```
Comment (pfrankov) [inline src/monitor.ts:297]: `fetchedMessage.attachments` comes directly from the VK API, but this extractor expects hydrated/transformed `vk-io` attachment objects. For location + media, this can silently lose the attachment. The raw direct path also bypasses `vk-io` text unescaping and reply/payload normalization. Prefer `context.loadMessagePayload()` and consume the normalized context.

File: src/monitor.ts [hunk 6]
```diff
@@ -112,20 +320,31 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
           ? context.conversationMessageId
-          : undefined,
+          : typeof fetchedMessage?.conversation_message_id === "number" &&
+              Number.isFinite(fetchedMessage.conversation_message_id)
+            ? fetchedMessage.conversation_message_id
+            : undefined,
       peerId,
       senderId,
-      text,
+      text: visibleBody,
```
Comment (pfrankov) [inline src/monitor.ts:332]: Keep provider text raw in the monitor. `handleVkInbound` already owns body construction. Carry a standard `NormalizedLocation`, render it once with `formatLocationText`, and spread `toLocationContext` into the final context. Otherwise hooks receive no structured location and adding coordinates here can alter command parsing.

File: src/monitor.ts [hunk 7]
```diff
@@ -112,20 +320,31 @@ export async function monitorVkProvider(opts: VkMonitorOptions): Promise<void> {
...
+      messagePayload: resolveVkMessagePayload(fetchedMessage?.payload ?? context.messagePayload),
+      geo,
       attachments,
       replyToMessageId: replyContext.replyToMessageId,
       replyToText: replyContext.replyToText,
     };
 
+    if ((context.hasGeo || rawGeo || fetchedMessage?.geo) && !geo) {
+      opts.runtime.log?.(
+        `vk: geo present but unresolved for peerId=${peerId} messageId=${context.id}; rawGeo=${JSON.stringify(rawGeo)}`,
```
Comment (pfrankov) [inline src/monitor.ts:344]: Do not write `rawGeo` to logs: it contains precise personal location data, and `JSON.stringify` on an unexpected object can itself throw. Log only a debug/verbose reason or schema kind, without coordinates, place text, peer ID, or message ID.

File: src/media.ts [hunk 8]
```diff
@@ -448,6 +448,80 @@ export function resolveVkInboundReplyContext(replyMessage: unknown): {
...
+    typeof record.latitude === "number" &&
+    Number.isFinite(record.latitude) &&
+    typeof record.longitude === "number" &&
+    Number.isFinite(record.longitude)
+  ) {
+    latitude = record.latitude;
+    longitude = record.longitude;
+  }
+
+  if (latitude === undefined || longitude === undefined) {
```
Comment (pfrankov) [inline src/media.ts:496]: Before accepting the result, validate latitude within `[-90, 90]` and longitude within `[-180, 180]`. The string branch should require exactly two components rather than accepting extra tokens. Please use OpenClaw's `NormalizedLocation` contract and add boundary/malformed-input tests.

File: src/media.ts [hunk 9]
```diff
@@ -448,6 +448,80 @@ export function resolveVkInboundReplyContext(replyMessage: unknown): {
...
+export function formatVkInboundGeoText(geo: VkInboundGeo | undefined): string {
+  if (!geo) {
+    return "";
+  }
+
+  const placeParts = [geo.placeTitle?.trim(), geo.city?.trim()].filter(
+    (entry): entry is string => Boolean(entry),
+  );
+  const placeText = placeParts.length > 0 ? ` (${placeParts.join(", ")})` : "";
+  return `[VK location] ${geo.latitude}, ${geo.longitude}${placeText}`;
```
Comment (pfrankov) [inline src/media.ts:522]: OpenClaw 2026.8.1 already provides `formatLocationText` and `toLocationContext`. Use those instead of a provider-specific formatter. The standard formatter deliberately keeps provider labels/addresses out of inline safe text and carries them as structured metadata; it also gives deterministic coordinate precision.

**Review by pfrankov — CHANGES_REQUESTED** [review]
I re-reviewed this end to end against the current plugin `main` and OpenClaw 2026.8.1. The two previous review points were addressed correctly, and I resolved those threads.

The feature itself is still needed: the current VK path can still turn a native location-only update into an empty inbound body. However, this branch is now three commits behind, has merge conflicts, has no current CI run, and the implementation should be redesigned around the location contract that now exists in OpenClaw rather than merged as-is.

## 1. Use OpenClaw's standard location contract

OpenClaw 2026.8.1 exports `NormalizedLocation`, `formatLocationText`, and `toLocationContext` from `openclaw/plugin-sdk/channel-inbound`.

Please use that contract instead of introducing a parallel `VkInboundGeo` representation and custom formatter:

- carry `location?: NormalizedLocation` in `VkInboundMessage`;
- let `monitor.ts` extract/normalize provider data only;
- let `handleVkInbound` remain the single owner of visible-body construction;
- render the safe inline location with `formatLocationText`;
- spread `toLocationContext(location)` into the finalized inbound context.

At the moment `message.geo` is populated but never consumed by `handleVkInbound`. The only working path is that `monitor.ts` pre-renders the location into `message.text`. Consequently:

- OpenClaw message hooks do not receive structured `event.location`;
- session/context metadata do not contain `LocationLat`, `LocationLon`, `LocationName`, etc.;
- the monitor and inbound layer both become responsible for presentation;
- appending coordinates to `message.text` can change slash/control-command parsing.

Keep command detection and `CommandBody` based on the original text or callback payload, while `BodyForAgent`/`RawBody` may include the formatted location.

OpenClaw's standard formatter also intentionally keeps provider-supplied place labels/addresses out of the safe inline text and carries them as structured untrusted metadata. `place.title` and `city` should not be concatenated directly into the body.

## 2. Do not bypass `vk-io` hydration

For polling contexts, `vk-io` already exposes the public `loadMessagePayload()` method. It resolves the exact message through `getById` or `getByConversationMessageId`, then runs `applyPayload`, which:

- unescapes message text;
- transforms raw VK attachments into `vk-io` attachment objects;
- normalizes reply messages;
- parses message payloads;
- updates the context's filled state.

This PR first performs the same API request manually and normally consumes the raw API object instead. That bypasses the normalization above. In particular, raw VK API attachments are nested (`{ type: "photo", photo: {...} }`), while `extractVkInboundAttachments` expects the hydrated `vk-io` attachment shape/getters. A location + photo/document message can therefore preserve the coordinates but lose its media.

The intended sequence should be:

1. Try to parse a complete geo object already present in the Bots Long Poll payload. If it is usable, perform zero API calls.
2. If there is only the User Long Poll geo marker/placeholder, call `context.loadMessagePayload()` once.
3. Re-read the hydrated `context.geo`, `context.attachments`, `context.replyMessage`, and `context.messagePayload`.
4. Only if exact hydration genuinely cannot resolve the location, fail softly or try another exact-ID endpoint.

Do not inspect the protected/internal `$filled` property. It is unnecessary and couples the plugin to `vk-io` internals.

There is also a concrete control-flow bug: when the manual direct fetch returns no item and `loadMessagePayload()` succeeds, `fetchedMessage` remains `undefined`. The next block therefore still calls `getHistory`, even though the context now contains the hydrated geo.

## 3. Remove fuzzy history matching

The `getHistory` fallback must not attach a location based on matching sender plus a ±120-second timestamp window. In an active DM or group, this can select a different message from the same sender and attribute the wrong coordinates to the current turn. Location is sensitive data; this path must fail closed rather than guess.

Use only stable exact identity:

- nonzero global message ID; or
- `peer_id` + `conversation_message_id`.

If both exact lookups fail, leave location unresolved. If a history fallback is retained despite that, it must require one uniquely matching candidate with exact identifiers; text/sender/time heuristics are not sufficient.

Also fetch only the missing location. Do not replace the event's text, attachments, callback payload, and reply context with values from a follow-up/history item. A delayed API read can observe an edited message, and a mistaken history match can mix another message's content with the current sender and message identity.

## 4. Validate and format coordinates defensively

`Number.isFinite` is not enough. Please enforce:

- latitude in `[-90, 90]`;
- longitude in `[-180, 180]`;
- exactly two components for the string form;
- rejection of extra/ambiguous tokens instead of silently taking the first two;
- deterministic formatting, preferably the standard six-decimal OpenClaw formatter.

Add tests for zero coordinates, boundaries, swapped/out-of-range values, extra tokens, malformed decimal separators, and non-finite values.

## 5. Do not log location payloads

The unresolved path writes `JSON.stringify(rawGeo)` together with peer and message IDs. Exact coordinates are sensitive personal data and should not enter normal logs. `JSON.stringify` can also throw on an unexpected/circular object and drop the inbound event.

Remove raw geo, full place data, and identifiers from this diagnostic. A verbose/debug-only message with a safe schema/reason code is sufficient. The unconditional per-message diagnostic should also be removed or placed behind the repository's normal verbose logging facility.

## 6. Preserve stable message identity

When `context.id === 0`, the exact conversation-message lookup may return a nonzero `fetchedMessage.id`, but the constructed message still uses `messageId: "0"`. That can collapse multiple updates onto the same identity and makes `markMessageReadVk` operate on zero. Use the resolved global ID when available, or define a stable provider identity based on peer + conversation message ID. Likewise, use the resolved message timestamp as a fallback when the polling context lacks a valid one.

## 7. Rebase and current compatibility checks

Please rebase onto the current `main` and resolve the inbound/media changes against the OpenClaw 2026.8.1 migration. In particular:

- use focused SDK imports; the broad `openclaw/plugin-sdk` import is rejected by the current compatibility guard;
- preserve the current Node 22 target, package metadata, and OpenClaw peer/compat constraints;
- run `npm run check` and the current OpenClaw 2026.8.1 / latest / beta runtime matrix.

## Required end-to-end coverage

The current tests stop at the object passed to `handleVkInbound`; they do not prove that location survives the complete OpenClaw pipeline. Please add coverage from a realistic `message_new` context through final inbound-context construction and hook projection:

1. Bots Long Poll with a complete raw geo object: no VK API enrichment call.
2. User Long Poll with the `{}` geo marker: exactly one successful hydration, then no history call.
3. Direct-fetch failure followed by successful `loadMessagePayload`: no history call.
4. Location + photo/document: both structured location and media facts survive.
5. Location-only: non-empty `BodyForAgent`/`RawBody`, plus `LocationLat` and `LocationLon`.
6. Text + location: text and location survive; command parsing still sees the original command body.
7. Final `message_received`/inbound hook event contains structured `event.location`.
8. Missing exact identity or ambiguous history candidates never attach another message's location.
9. Invalid/out-of-range coordinates are rejected without logging the raw payload.
10. Preserve `conversationMessageId`, stable `messageId`, reply context, and callback payload across hydration.

Please base the parser tests on sanitized real fixtures for both Bots Long Poll and User Long Poll/API hydration, including the claimed `"lon lat"` string shape. The change is useful, but the current implementation is not safe to merge.

**Review by pfrankov — CHANGES_REQUESTED** [review]
Re-reviewed head `5eb12a0`. The earlier fixes for ordinary text-message enrichment and preserving `conversationMessageId` are present. However, the blocking issues raised in the September 1 review are still in the current code, and GitHub reports merge conflicts.

Please address the following before merging:

1. **[P1] Match location to an exact message, never by similarity.** `fetchVkApiMessagePayloadFromHistory()` still accepts matching text/sender or a same-sender timestamp within 120 seconds. This can attach a different message's location (and, through the subsequent merge, other message fields) to the current turn. Use only a valid exact message ID or exact peer + valid conversation message ID. Do not let missing/undefined conversation IDs compare equal. If there is no exact match, leave geo unresolved. Add tests with two nearby messages from the same sender, identical text, and missing conversation IDs.

2. **[P1] Preserve the normalized vk-io message and avoid redundant history queries.** The direct API/history path feeds raw `fetchedMessage.attachments` into an extractor that expects hydrated vk-io attachment objects. Location + media can therefore lose attachments. Also, successful `context.loadMessagePayload()` leaves `fetchedMessage` undefined, so `!fetchedMessage?.geo` still triggers a history query even if hydration recovered the location. Read available context geo first; hydrate once only when required; re-read the normalized context before considering any exact-ID fallback. Keep raw API geo recovery separate from text/attachment/reply normalization. Cover location + photo/document/reply and successful hydration with no extra history fetch.

3. **[P1] Remove location data from logs.** The unresolved-geo path still logs `JSON.stringify(rawGeo)`, including precise coordinates/place information, and serialization itself can throw before dispatch. Log a bounded reason/schema kind only, without raw geo or personal identifiers.

4. **Keep body construction in the inbound layer.** The monitor currently stores the rendered location in `message.text`. Preserve the provider's original text and carry location separately; render it once in inbound and populate the SDK's structured location context, as requested in the existing review. Add a command + location regression so coordinates do not alter command parsing.

Finally, rebase onto current `main`, resolve the conflicts without dropping the newer lifecycle/readiness changes, and run the full project checks on the rebased head. I did not find a PR-triggered CI run for this head; the old reported test count is not validation against current main.
